File: java/79_WordSearch.java
class Solution {
    int n, m;
    public boolean exist(char[][] board, String word) {
        n = board.length;
        m = board[0].length;

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (board[i][j] == word.charAt(0) && bfs(i, j, 0, board, word)) {
                    return true;
                }
            }
        }
        return false;
    }

    private boolean check(int i, int j, int index, char[][] board, String word) {
        if (i < 0 || i >= n || j < 0 || j >= m || board[i][j] != word.charAt(index)) {
            return false;
        }
        return true;
    }

    private boolean bfs(int i, int j, int index, char[][] board, String word) {
        if (index == word.length()) {
            return true;
        }
        if (!check(i, j, index, board, word)) {
            return false;
        }
        char temp = board[i][j];
        board[i][j] = ' ';
        boolean res = bfs(i-1, j, index+1, board, word) ||
                      bfs(i+1, j, index+1, board, word) ||
                      bfs(i, j+1, index+1, board, word) ||
                      bfs(i, j-1, index+1, board, word);
        board[i][j] = temp;
        return res;
    }
}

File: c++/79_WordSearch.cpp
class Solution {
public:
    bool exist(vector<vector<char>>& board, string word) {
        row = (int)board.size();
        col = (int)board[0].size();
        for (int i = 0; i < row; i++) {
            for (int j = 0; j < col; j++) {
                if (dfs(board, word, i, j, 0)) {
                    return true;
                }
            }
        }
        return false;
    }
    
private:
    int row, col;
    
    bool dfs(vector<vector<char>>& board, string word, int i, int j, int k) {
        if (i >= row || i < 0 || j >= col || j < 0 || board[i][j] != word[k]) {
            return false;
        }
        if (k == word.length() - 1) return true;
        
        board[i][j] = '\0'; //visited
        
        bool res = dfs(board, word, i+1, j, k+1) ||
                   dfs(board, word, i-1, j, k+1) ||
                   dfs(board, word, i, j+1, k+1) ||
                   dfs(board, word, i, j-1, k+1);
        
        board[i][j] = word[k];
        return res;
    }
};